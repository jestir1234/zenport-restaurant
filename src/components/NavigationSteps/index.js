import React from "react";
import PropTypes from "prop-types";
import styled from "react-emotion";
import { MEDIA_BREAKPOINTS, COLORS } from "../../constants";

const Container = styled("div")(props => ({
  display: "flex",
  justifyContent: "center",
  padding: "30px 5px 30px 5px"
}));

const NavContainer = styled("div")(props => ({
  display: "flex",
  justifyContent: "center"
}));

const NavItem = styled("div")(props => ({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  whiteSpace: "nowrap",
  padding: "20px",
  margin: "0 10px 0 10px",
  border: `1px solid ${COLORS.darkAqua}`,
  borderRadius: "10%",
  [MEDIA_BREAKPOINTS.down("xs")]: {
    padding: "10px",
    margin: "0 5px 0 5px"
  },
  background: props.currentPage === props.step ? COLORS.darkAqua : COLORS.white,
  color: props.currentPage === props.step ? COLORS.white : COLORS.dark,
  fontWeight: props.currentPage === props.step ? "bold" : "300",
  boxShadow:
    props.currentPage === props.step
      ? "6px 10px 14px -4px rgba(0,0,0,0.75)"
      : "none"
}));

NavItem.propTypes = {
  currentPage: PropTypes.string
};

const STEPS = 4;

const NavigationSteps = props => {
  return (
    <Container>
      <NavContainer>
        {[...Array(STEPS)].map((step, idx) => (
          <NavItem
            currentPage={props.currentPage}
            key={idx}
            step={`step${idx + 1}`}
          >
            Step {idx + 1}
          </NavItem>
        ))}
      </NavContainer>
    </Container>
  );
};

NavigationSteps.propTypes = {
  currentPage: PropTypes.string
};

export default NavigationSteps;
