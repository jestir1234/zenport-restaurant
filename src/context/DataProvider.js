import React, { Component } from "react";
import data from "../data.json";

export const DataContext = React.createContext();

const parseRestaurants = () => {
  let restaurants = {};
  let uniqRestaurants = Array.from(
    new Set(data[0].dishes.map(dish => dish.restaurant))
  );
  uniqRestaurants.forEach(name => {
    restaurants[name] = { availableMeals: "", menu: [] };
  });

  data[0].dishes.forEach(dish => {
    let currentMenu = restaurants[dish.restaurant].menu;
    currentMenu.push(dish.name);
    restaurants[dish.restaurant] = {
      availableMeals: dish.availableMeals,
      menu: currentMenu
    };
  });
  return restaurants;
};

const parseMealTypes = restaurants => {
  let mealTypes = {};
  Object.keys(restaurants).forEach(restaurant => {
    let restaurantMealTypes = restaurants[restaurant]["availableMeals"];
    restaurantMealTypes.forEach(type => {
      if (mealTypes[type]) {
        mealTypes[type]["restaurants"][restaurant] = restaurants[restaurant];
      } else {
        mealTypes[type] = { restaurants: {} };
      }
    });
  });
  return mealTypes;
};

const defaultStepOptions = () => ({
  step1: {
    selectedMealType: null,
    selectedPartyCount: { value: 1, label: 1 }
  },
  step2: {
    selectedRestaurant: null
  },
  step3: {
    selectedMeal: null,
    selectedServings: { value: 1, label: 1 },
    currentOrder: {},
    validMealCount: false,
    errorMessage: ""
  }
});

class DataProvider extends Component {
  state = {
    mealTypes: parseMealTypes(parseRestaurants()),
    stepOptions: defaultStepOptions()
  };

  handleChange = (selectedOption, option, currentStep) => {
    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        [currentStep]: {
          ...prevState.stepOptions[currentStep],
          [option]: selectedOption
        }
      }
    }));
  };

  addOrder = order => {
    let newOrder = { ...this.state.stepOptions.step3.currentOrder, ...order };
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder: newOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  subtractOrder = dish => {
    let currentOrder = { ...this.state.stepOptions.step3.currentOrder };
    delete currentOrder[dish];
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  validateMealCount = () => {
    let currentOrders = this.state.stepOptions.step3.currentOrder;
    let totalMeals =
      Object.keys(currentOrders).length > 0
        ? Object.values(this.state.stepOptions.step3.currentOrder).reduce(
            (acc, val) => acc + val
          )
        : 0;

    let selectedPartyCount = this.state.stepOptions.step1.selectedPartyCount
      .value;
    let errorMessage = "";

    if (totalMeals > 10) {
      errorMessage = "Exceeded max total servings of 10.";
    }

    if (totalMeals < selectedPartyCount) {
      errorMessage =
        "Total servings must be equal to or greater than the no. of people in your group.";
    }

    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        step3: {
          ...prevState.stepOptions.step3,
          validMealCount: totalMeals <= 10 && totalMeals >= selectedPartyCount,
          errorMessage
        }
      }
    }));
  };

  render() {
    return (
      <DataContext.Provider
        value={{
          state: this.state,
          handleChange: (selectedOption, option, currentStep) =>
            this.handleChange(selectedOption, option, currentStep),
          addOrder: order => this.addOrder(order),
          subtractOrder: dish => this.subtractOrder(dish)
        }}
      >
        {this.props.children}
      </DataContext.Provider>
    );
  }
}

export default DataProvider;
