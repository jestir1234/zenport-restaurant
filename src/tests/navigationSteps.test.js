import React from "react";
import NavigationSteps from "../components/NavigationSteps";
import renderer from "react-test-renderer";
import DataProvider from "../context/DataProvider";

const currentPage = {
  next: "step2",
  page: "step1",
  prev: "step1"
};

test("Renders the <NavigationSteps /> component", () => {
  const component = renderer.create(
    <DataProvider>
      <NavigationSteps currentPage={currentPage.page} />
    </DataProvider>
  );
  let tree = component.toJSON();
  expect(tree);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
