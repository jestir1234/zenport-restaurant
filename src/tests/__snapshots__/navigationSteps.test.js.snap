// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the <NavigationSteps /> component 1`] = `
<div
  className="css-ddmbx9"
>
  <div
    className="css-1l4w6pd"
  >
    <div
      className="css-1mlsixv"
      step="step1"
    >
      Step 
      1
    </div>
    <div
      className="css-fexx03"
      step="step2"
    >
      Step 
      2
    </div>
    <div
      className="css-fexx03"
      step="step3"
    >
      Step 
      3
    </div>
    <div
      className="css-fexx03"
      step="step4"
    >
      Step 
      4
    </div>
  </div>
</div>
`;
